src/main/java/at/hannibal2/skyhanni/features/skillprogress/SkillProgress.kt

```diff
@@ -0,0 +1,458 @@
+        var targetLevel = if (useCustomGoalLevel) skillInfo.customGoalLevel else level + 1
+        if (targetLevel <= level || targetLevel > 400) targetLevel = (level + 1)
+        val currentLevelNeededXp = SkillUtil.xpRequiredForLevel(level.toDouble()) + skillInfo.overflowCurrentXp
+        val targetNeededXp = SkillUtil.xpRequiredForLevel(targetLevel.toDouble())
+        var remaining = if (useCustomGoalLevel) targetNeededXp - currentLevelNeededXp else skillInfo.overflowCurrentXpMax - skillInfo.overflowCurrentXp
+
+        if (!useCustomGoalLevel) {
+            if (skillInfo.overflowCurrentXpMax == skillInfoLast.overflowCurrentXpMax) {
+                remaining = interpolate(remaining.toFloat(), (skillInfoLast.overflowCurrentXpMax - skillInfoLast.overflowCurrentXp).toFloat(), lastGainUpdate.toMillis()).toLong()
+            }
+        }
+
+        add(Renderable.string(buildString {
+            append("§6Skill: §7${activeSkill.displayName} ")
+            if (useCustomGoalLevel && targetLevel != 0) append("§8$level➜§3${targetLevel}") else append(level)
+        }))
+        if (useCustomGoalLevel)
+            add(Renderable.string("§7Needed XP: §e${remaining.addSeparators()}"))
+
+        var xpInterp = xpInfo.xpGainHour
+        if (xpInfo.xpGainHour < 1000) {
+            add(Renderable.string("§7In §cN/A"))
+        } else {
+            add(Renderable.string("§7In §b${Utils.prettyTime((remaining) * 1000 * 60 * 60 / xpInterp.toLong())} " +
+                if (xpInfo.isActive) "" else "§c(PAUSED)"))
+        }
+
+        if (xpInfo.xpGainLast == xpInfo.xpGainHour && xpInfo.xpGainHour <= 0) {
+            add(Renderable.string("§7XP/h: §cN/A"))
+        } else {
+            xpInterp = interpolate(xpInfo.xpGainHour, xpInfo.xpGainLast, lastGainUpdate.toMillis())
+            add(Renderable.string("§7XP/h: §e${xpInterp.toLong().addSeparators()} " +
+                if (xpInfo.isActive) "" else "§c(PAUSED)"))
+        }
+
+
+        val session = xpInfo.timeActive.seconds.format(TimeUnit.HOUR)
+        add(Renderable.clickAndHover("§7Session: §e$session ${if (xpInfo.sessionTimerActive) "" else "§c(PAUSED)"}",
+            listOf("§eClick to reset!")) {
+            xpInfo.sessionTimerActive = false
+            xpInfo.shouldStartTimer = true
+            xpInfo.timeActive = 0L
+            chat("Timer for §b${activeSkill.displayName} §ehas been reset!")
+        })
+    }
+
+    private fun drawDisplay() = buildList {
+        val skillMap = SkillAPI.storage?.skillData ?: return@buildList
+        val skill = skillMap[activeSkill] ?: return@buildList
+        val useCustomGoalLevel = skill.customGoalLevel != 0 && skill.customGoalLevel > skill.overflowLevel
+        val targetLevel = skill.customGoalLevel
+        val xp = skill.totalXp
+        val currentLevel = if (xp <= 111672425L) {
```

Comment:
111672425L is a number I also saw in SkillAPI, maybe make it a constant there with a descriptive name so i know what it is :)